codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java

```diff
@@ -0,0 +1,318 @@
+            // associated with the stream, as sending a RST_STREAM means that someone "doesn't care"
+            // about the stream anymore and thus there is not point in failing the promises and invoking
+            // error handling routines.
+            stream.close(null);
+            promise.setSuccess();
+        } else {
+            promise.setFailure(
+                    connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+        }
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data,
+                                   int padding,
+                                   boolean endOfStream, ChannelPromise promise) {
+        if (existingStream(streamId)) {
+            return super.writeData(ctx, streamId, data, padding, endOfStream, promise);
+        }
+        PendingStream pendingStream = pendingStreams.get(streamId);
+        if (pendingStream != null) {
+            pendingStream.frames.add(new DataFrame(data, padding, endOfStream, promise));
+        } else {
+            ReferenceCountUtil.safeRelease(data);
+            promise.setFailure(
+                    connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+        }
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+        receivedSettings = true;
+        ChannelFuture future = super.writeSettingsAck(ctx, promise);
+        // After having received a SETTINGS frame, the maximum number of concurrent streams
+        // might have changed. So try to create some buffered streams.
+        tryCreatePendingStreams();
+        return future;
+    }
+
+    @Override
+    public void close() {
+        super.close();
+        cancelPendingStreams();
+    }
+
+    private void tryCreatePendingStreams() {
+        while (!pendingStreams.isEmpty() && canCreateStream()) {
+            Map.Entry<Integer, PendingStream> entry = pendingStreams.pollFirstEntry();
+            PendingStream pendingStream = entry.getValue();
+            pendingStream.sendFrames();
+            largestCreatedStreamId = pendingStream.streamId;
+        }
+    }
+
+    private void cancelPendingStreams() {
+        Exception e = new Exception("Connection closed.");
+        while (!pendingStreams.isEmpty()) {
+            PendingStream stream = pendingStreams.pollFirstEntry().getValue();
+            stream.close(e);
+        }
+    }
+
+    private void cancelGoAwayStreams(int lastStreamId) {
+        Iterator<PendingStream> iter = pendingStreams.values().iterator();
+        while (iter.hasNext()) {
+            PendingStream stream = iter.next();
+            if (stream.streamId > lastStreamId) {
+                iter.remove();
+                stream.close(Http2Exception.streamError(stream.streamId, Http2Error.STREAM_CLOSED,
```

Comment:
@nmittler so `goAwayStatus` would never be called for those buffered streams, because they only exist in the buffering encoder. So I think good call to keep the custom exception type. I think it would be useful to document somewhere that this exception exists and will be set. WDYT?